plugins/ndpi/ndpi.c

```diff
@@ -0,0 +1,537 @@
+
+        NDPI_ZERO_BIT(risk_mask);
+
+        if (dup != NULL) {
+            token = strtok_r(dup, ",", &tmp);
+
+            while (token != NULL) {
+                ndpi_risk_enum risk_id = ndpi_code2risk(token);
+                if (risk_id >= NDPI_MAX_RISK) {
+                    SCLogError("unrecognized risk '%s', "
+                               "please check ndpiReader -H for valid risk codes",
+                            token);
+                    return NULL;
+                }
+                NDPI_SET_BIT(risk_mask, risk_id);
+                token = strtok_r(NULL, ",", &tmp);
+            }
+
+            SCFree(dup);
+        }
+    }
+
+    data = SCMalloc(sizeof(DetectnDPIRiskData));
+    if (unlikely(data == NULL))
+        return NULL;
+
+    data->risk_mask = risk_mask;
+    data->negated = negate;
+
+    return data;
+}
+
+static bool nDPIRiskDataHasConflicts(const DetectnDPIRiskData *us, const DetectnDPIRiskData *them)
+{
+    /* check for duplicate */
+    if (us->risk_mask == them->risk_mask)
+        return true;
+
+    return false;
+}
+
+static int DetectnDPIRiskSetup(DetectEngineCtx *de_ctx, Signature *s, const char *arg)
+{
+    DetectnDPIRiskData *data = DetectnDPIRiskParse(arg, s->init_data->negated);
+    if (data == NULL)
+        goto error;
+
+    SigMatch *tsm = s->init_data->smlists[DETECT_SM_LIST_MATCH];
+    for (; tsm != NULL; tsm = tsm->next) {
+        if (tsm->type == ndpi_risk_keyword_id) {
+            const DetectnDPIRiskData *them = (const DetectnDPIRiskData *)tsm->ctx;
+
+            if (nDPIRiskDataHasConflicts(data, them)) {
+                SCLogError("can't mix "
+                           "positive ndpi-risk match with negated");
+                goto error;
+            }
+        }
+    }
+
+    if (SigMatchAppendSMToList(de_ctx, s, ndpi_risk_keyword_id, (SigMatchCtx *)data,
+                DETECT_SM_LIST_MATCH) == NULL) {
+        goto error;
+    }
+    return 0;
+
+error:
+    if (data != NULL)
+        SCFree(data);
+    return -1;
+}
+
+static void DetectnDPIRiskFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCFree(ptr);
+}
+
+static void EveCallback(ThreadVars *tv, const Packet *p, Flow *f, JsonBuilder *jb, void *data)
+{
+    /* Adding ndpi info to EVE requires a flow. */
+    if (f == NULL) {
+        return;
+    }
+
+    struct NdpiThreadContext *threadctx = ThreadGetStorageById(tv, thread_storage_id);
+    struct NdpiFlowContext *flowctx = FlowGetStorageById(f, flow_storage_id);
+    ndpi_serializer serializer;
+    char *buffer;
+    uint32_t buffer_len;
+
+    SCLogDebug("EveCallback: tv=%p, p=%p, f=%p", tv, p, f);
+
+    ndpi_init_serializer(&serializer, ndpi_serialization_format_inner_json);
+
+    /* Use ndpi_dpi2json to get a JSON with nDPI metadata */
+    ndpi_dpi2json(threadctx->ndpi, flowctx->ndpi_flow, flowctx->detected_l7_protocol, &serializer);
+
+    buffer = ndpi_serializer_get_buffer(&serializer, &buffer_len);
```

Comment:
Can `ndpi_serializer_get_buffer` return `NULL`?

Reply:
Appears not.